src/ripple/nodestore/impl/ShardInfo.cpp

```diff
@@ -0,0 +1,124 @@
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/main/Application.h>
+#include <ripple/nodestore/ShardInfo.h>
+#include <ripple/protocol/digest.h>
+
+#include <boost/algorithm/clamp.hpp>
+
+namespace ripple {
+namespace NodeStore {
+
+std::string
+ShardInfo::finalizedToString() const
+{
+    if (!finalized_.empty())
+        return ripple::to_string(finalized_);
+    return {};
+}
+
+std::string
+ShardInfo::incompleteToString() const
+{
+    std::string result;
+    if (!incomplete_.empty())
+    {
+        for (auto const& [shardIndex, incomplete] : incomplete_)
+        {
+            result += std::to_string(shardIndex) + ":" +
+                std::to_string(incomplete.percentProgress) + ",";
+        }
+        result.pop_back();
+    }
+
+    return result;
+}
+
+bool
+ShardInfo::update(
+    std::uint32_t shardIndex,
+    ShardState state,
+    std::uint32_t percentProgress)
+{
+    if (state == ShardState::finalized)
+    {
+        if (boost::icl::contains(finalized_, shardIndex))
+            return false;
+
+        finalized_.insert(shardIndex);
+        return true;
+    }
+
+    return incomplete_.emplace(shardIndex, Incomplete(state, percentProgress))
+        .second;
+}
+
+protocol::TMPeerShardInfoV2
+ShardInfo::makeMessage(Application& app) const
+{
+    protocol::TMPeerShardInfoV2 message;
+    Serializer s;
+    s.add32(HashPrefix::shardInfo);
+
+    if (!incomplete_.empty())
+    {
+        message.mutable_incomplete()->Reserve(incomplete_.size());
+        for (auto const& [shardIndex, incomplete] : incomplete_)
+        {
+            auto tmIncomplete{message.add_incomplete()};
+
+            tmIncomplete->set_shardindex(shardIndex);
+            s.add32(shardIndex);
+
+            auto const state{static_cast<std::uint32_t>(incomplete.state)};
+            tmIncomplete->set_state(state);
+            s.add32(state);
+
+            // Set progress if greater than zero
+            if (incomplete.percentProgress > 0)
+            {
+                tmIncomplete->set_progress(incomplete.percentProgress);
+                s.add32(incomplete.percentProgress);
+            }
+        }
+    }
+
+    if (!finalized_.empty())
+    {
+        auto const str{ripple::to_string(finalized_)};
+        message.set_finalized(str);
+        s.addRaw(str.data(), str.size());
+    }
+
+    // Set the public key
+    auto const& publicKey{app.nodeIdentity().first};
+    message.set_publickey(publicKey.data(), publicKey.size());
+
+    // Create a digital signature using the node private key
+    auto const signature{sign(publicKey, app.nodeIdentity().second, s.slice())};
```

Comment:
Maybe we should add a timestamp to ShardInfo. I don't have a strong reason, but three minor ones:
- For message freshness, so that an older message cannot be replayed. Not sure if possible, but imagine an attacker collects old messages (with valid signatures) and send them out later. A node might be tricked to use an old shardInfo to replace a new shardInfo,  https://github.com/miguelportilla/rippled/blob/2a0b097178d61c16f6e09b57cbdce3734becd0ea/src/ripple/overlay/impl/PeerImp.cpp#L1365
- In `crawlShards()` the `ShardInfo`s are combined, if there are two `ShardInfo`s of the same node, the newer one can win. https://github.com/miguelportilla/rippled/blob/2a0b097178d61c16f6e09b57cbdce3734becd0ea/src/ripple/overlay/impl/OverlayImpl.cpp#L777
- `Serializer s;` is signed. But the mapping from `ShardInfo` to `Serializer` is not 1:1. E.g. a `ShardInfo` with 2 `incomplete` with `percentProgress > 0` and a `ShardInfo` with 3 `incomplete` with `percentProgress == 0`. There is a small chance that their `Serializer`s are the same. Then the signature on one can be put on the other which is never signed. Adding the timestamp right after `s.add32(HashPrefix::shardInfo)` can solve the problem.

Reply:
Great idea, @pwang200, thanks!